### PR TITLE
(BIDS-2279) Add head parameter to slot endpoint

### PR DESCRIPTION
**Pull request by D13ce**
This PR adds the possibility to use "head" as parameter for the API V1 slot endpoint. When used, it will return the slot containing the chain's head block.
Swagger has been updated accordingly.

### CODE REVIEW

**Review by Eisei24 — APPROVED** [review]
lgtm

I could not reproduce the bug before or after the changes.

File: handlers/api.go [hunk 1]
```diff
@@ -372,37 +372,54 @@ func ApiEpochSlots(w http.ResponseWriter, r *http.Request) {
...
-		}
-	}
+	blockRootHash := []byte{}
 
 	if slotOrHash == "latest" {
+		// simply check the latest slot (might be empty which causes an error)
 		blockSlot = int64(services.LatestSlot())
+	} else if slotOrHash == "head" {
+		// retrieve the slot containing the head block of the chain (cannot return an empty slot unless there is no block in the chain yet)
+		err := db.ReaderDb.Get(&blockRootHash, `
```
Comment (peterbitfly) [inline handlers/api.go:397]: this should not hit the database, instead create a new property in the services module that gets updated via the FDU

**Review by peterbitfly — CHANGES_REQUESTED** [review]

**Review by Eisei24 — APPROVED** [review]
lgtm

File: services/execution_layer.go [hunk 2]
```diff
@@ -34,14 +36,60 @@ func latestBlockUpdater(wg *sync.WaitGroup) {
...
+		blockRootHash := []byte{}
+		err := db.ReaderDb.Get(&blockRootHash, `
+		SELECT blockroot
+		FROM blocks
+		WHERE status = '1'
+		ORDER BY slot DESC
+		LIMIT 1`)
+
+		if err != nil {
+			logger.WithError(err).Error("error getting blockrroot hash for chain head")
```
[the remark this answers is not in the document]
Comment (D13ce) [reply services/execution_layer.go]: Fixed.

File: services/execution_layer.go [hunk 3]
```diff
@@ -34,14 +36,60 @@ func latestBlockUpdater(wg *sync.WaitGroup) {
...
+// headBlockRootHashUpdater updates the hash of the current chain head block
+func headBlockRootHashUpdater(wg *sync.WaitGroup) {
+	firstRun := true
+
+	for {
+		blockRootHash := []byte{}
+		err := db.ReaderDb.Get(&blockRootHash, `
+		SELECT blockroot
+		FROM blocks
+		WHERE status = '1'
```
Comment (recy21) [inline services/execution_layer.go:62]: There is no index for just `status` we should use one, or proof that a "multi-index" will cover this case.

File: services/execution_layer.go [hunk 4]
```diff
@@ -34,14 +36,60 @@ func latestBlockUpdater(wg *sync.WaitGroup) {
...
+		ORDER BY slot DESC
```
Comment (recy21) [inline services/execution_layer.go:63]: There is no index for just `slot` we should use one, or proof that a "multi-index" will cover this case.

File: services/execution_layer.go [hunk 5]
```diff
@@ -34,14 +36,60 @@ func latestBlockUpdater(wg *sync.WaitGroup) {
...
+		LIMIT 1`)
+
+		if err != nil {
+			logger.WithError(err).Error("error getting blockroot hash for chain head")
+		}
+		cacheKey := fmt.Sprintf("%d:frontend:%s", utils.Config.Chain.Config.DepositChainID, latestBlockHashRootCacheKey)
+		err = cache.TieredCache.SetString(cacheKey, string(blockRootHash), time.Hour*24)
+		if err != nil {
+			logger.Errorf("error caching %s: %v", latestBlockHashRootCacheKey, err)
```
Comment (recy21) [inline services/execution_layer.go:72]: Should use utils.LogError

File: services/execution_layer.go [hunk 6]
```diff
@@ -34,14 +36,60 @@ func latestBlockUpdater(wg *sync.WaitGroup) {
...
+		blockRootHash := []byte{}
+		err := db.ReaderDb.Get(&blockRootHash, `
+		SELECT blockroot
+		FROM blocks
+		WHERE status = '1'
+		ORDER BY slot DESC
+		LIMIT 1`)
+
+		if err != nil {
+			logger.WithError(err).Error("error getting blockroot hash for chain head")
```
Comment (recy21) [inline services/execution_layer.go:67]: Should use utils.LogError

File: services/execution_layer.go [hunk 7]
```diff
@@ -18,10 +20,10 @@ func latestBlockUpdater(wg *sync.WaitGroup) {
 		if err != nil {
 			logger.WithError(err).Error("error getting most recent eth1 block")
 		}
-		cacheKey := fmt.Sprintf("%d:frontend:latestEth1BlockNumber", utils.Config.Chain.Config.DepositChainID)
+		cacheKey := fmt.Sprintf("%d:frontend:%s", utils.Config.Chain.Config.DepositChainID, latestBlockNumberCacheKey)
 		err = cache.TieredCache.SetUint64(cacheKey, recent.GetNumber(), time.Hour*24)
 		if err != nil {
-			logger.Errorf("error caching latestBlockNumber: %v", err)
+			logger.Errorf("error caching %s: %v", latestBlockNumberCacheKey, err)
```
Comment (recy21) [inline services/execution_layer.go:26]: Should use utils.LogError

**Review by recy21 — CHANGES_REQUESTED** [review]

File: services/execution_layer.go (hunk 7 above)
Comment (D13ce) [reply services/execution_layer.go:26]: Fixed

File: services/execution_layer.go [hunk 8]
```diff
@@ -34,14 +36,60 @@ func latestBlockUpdater(wg *sync.WaitGroup) {
...
+// LatestEth1BlockNumber will return most recent eth1 block number
 func LatestEth1BlockNumber() uint64 {
-	cacheKey := fmt.Sprintf("%d:frontend:latestEth1BlockNumber", utils.Config.Chain.Config.DepositChainID)
+	cacheKey := fmt.Sprintf("%d:frontend:%s", utils.Config.Chain.Config.DepositChainID, latestBlockNumberCacheKey)
 
 	if wanted, err := cache.TieredCache.GetUint64WithLocalTimeout(cacheKey, time.Second*5); err == nil {
 		return wanted
 	} else {
-		logger.Errorf("error retrieving latestEth1BlockNumber from cache: %v", err)
+		logger.Errorf("error retrieving %s from cache: %v", latestBlockNumberCacheKey, err)
```
[the remark this answers is not in the document]
Comment (D13ce) [reply services/execution_layer.go:46]: Fixed

File: services/execution_layer.go [hunk 9]
```diff
@@ -9,6 +9,8 @@ import (
 	"time"
 )
 
+const latestBlockNumberCacheKey = "latestEth1BlockNumber"
```
[the remark this answers is not in the document]
Comment (D13ce) [reply services/execution_layer.go:12]: Moved

File: handlers/api.go (hunk 1 above)
Comment (D13ce) [reply handlers/api.go]: Added a new service accordingly

File: services/execution_layer.go (hunk 3 above)
Comment (D13ce) [reply services/execution_layer.go:62]: Here is your proof that the composite index `idx_blocks_status_slot_epoch` is used:

```
EXPLAIN ANALYZE
	SELECT blockroot
	FROM blocks
	WHERE status = '1'
	ORDER BY slot DESC
	LIMIT 1;
```
=>
```
... // some output
   ->  Index Scan Backward using idx_blocks_status_slot_epoch on blocks 
... // some output
```
This works as `idx_blocks_status_slot_epoch` has `status` as its first part, see here:
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/44141630c110fa8284724f18d5e6970fb92ab819/db/migrations/20230612084855_add_missing_indices.sql#L49

File: services/execution_layer.go (hunk 4 above)
Comment (D13ce) [reply services/execution_layer.go:63]: Similar to my answer here (https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2401#discussion_r1291084623), the already existing index `idx_blocks_status_slot_epoch` covers this as its second part is `slot`.
So this query ONLY uses `idx_blocks_status_slot_epoch` for both status and slot which works as the composite index has both columns as first and second part.

File: services/execution_layer.go (hunk 6 above)
Comment (D13ce) [reply services/execution_layer.go]: Fixed

File: services/execution_layer.go (hunk 5 above)
Comment (D13ce) [reply services/execution_layer.go]: Fixed

File: services/execution_layer.go (hunk 3 above)
Comment (recy21) [reply services/execution_layer.go:61]: good

File: services/execution_layer.go (hunk 4 above)
Comment (recy21) [reply services/execution_layer.go:62]: good

**Review by recy21 — APPROVED** [review]

**Review by peterbitfly — APPROVED** [review]

**Pull request merged**